### PR TITLE
all: make client use server-provided DERP map, add DERP region support

### LINKED ISSUE #387: Don't hard-code the DERP map
The DERP server map is currently hard-coded.

It should be dynamic and sent from the server. (This will permit us to reconfigure things over time, add servers more easily, support on-prem customers, etc.)

### LINKED ISSUE #399: wgengine: flaky tests, log in goroutine after test completed
At master (35a8586f7e7e6f6bddf57f7f0f09b16b3f4253fc) just now:

```
dev:tailscale.com $ go test -count=10 -v ./wgengine/
...
=== RUN   TestWatchdog
=== PAUSE TestWatchdog
=== CONT  TestWatchdog
=== RUN   TestWatchdog/default_watchdog_does_not_fire
=== PAUSE TestWatchdog/default_watchdog_does_not_fire
=== RUN   TestWatchdog/watchdog_fires_on_blocked_getStatus
=== PAUSE TestWatchdog/watchdog_fires_on_blocked_getStatus
=== CONT  TestWatchdog/default_watchdog_does_not_fire
=== CONT  TestWatchdog/watchdog_fires_on_blocked_getStatus
    TestWatchdog/watchdog_fires_on_blocked_getStatus: magicsock.go:300: magicsock: starting endpoint update (initial)
    TestWatchdog/watchdog_fires_on_blocked_getStatus: userspace.go:78: Routine: event worker - started
    TestWatchdog/default_watchdog_does_not_fire: magicsock.go:300: magicsock: starting endpoint update (initial)
    TestWatchdog/default_watchdog_does_not_fire: userspace.go:78: Routine: event worker - started
    TestWatchdog/watchdog_fires_on_blocked_getStatus: userspace.go:78: UDP bind has been updated
    TestWatchdog/watchdog_fires_on_blocked_getStatus: router_fake.go:24: Warning: fakeRouter.Up: not implemented.
    TestWatchdog/watchdog_fires_on_blocked_getStatus: router_fake.go:29: Warning: fakeRouter.Set: not implemented.
    TestWatchdog/default_watchdog_does_not_fire: userspace.go:78: UDP bind has been updated
    TestWatchdog/default_watchdog_does_not_fire: router_fake.go:24: Warning: fakeRouter.Up: not implemented.
    TestWatchdog/default_watchdog_does_not_fire: router_fake.go:29: Warning: fakeRouter.Set: not implemented.
    TestWatchdog/default_watchdog_does_not_fire: userspace.go:78: Device closing
    TestWatchdog/default_watchdog_does_not_fire: magicsock.go:505: magicsock.Conn.determineEndpoints: updateNetInfo: got STUN error: context canceled; missing replies from: derp1.tailscale.com:3478, derp2.tailscale.com:3478, derp3.tailscale.com:3478, derp4.tailscale.com:3478, derp5.tailscale.com:3478
    TestWatchdog/default_watchdog_does_not_fire: userspace.go:78: Routine: event worker - stopped
    TestWatchdog/default_watchdog_does_not_fire: magicsock.go:304: magicsock: endpoint update (initial) failed: got STUN error: context canceled; missing replies from: derp1.tailscale.com:3478, derp2.tailscale.com:3478, derp3.tailscale.com:3478, derp4.tailscale.com:3478, derp5.tailscale.com:3478
    TestWatchdog/default_watchdog_does_not_fire: userspace.go:78: Interface closed
    TestWatchdog/default_watchdog_does_not_fire: router_fake.go:34: Warning: fakeRouter.Close: not implemented.
panic: Log in goroutine after TestWatchdog/watchdog_fires_on_blocked_getStatus has completed

goroutine 34 [running]:
testing.(*common).logDepth(0xc0001c4900, 0xc00038c600, 0xf3, 0x3)
        /home/bradfitz/sdk/go1.14.2/src/testing/testing.go:680 +0x4d3
testing.(*common).log(...)
        /home/bradfitz/sdk/go1.14.2/src/testing/testing.go:662
testing.(*common).Logf(0xc0001c4900, 0xc00040e190, 0xc, 0xc000280070, 0x1, 0x1)
        /home/bradfitz/sdk/go1.14.2/src/testing/testing.go:701 +0x7e
tailscale.com/types/logger.WithPrefix.func1(0x8b443b, 0x2, 0xc000280070, 0x1, 0x1)
        /home/bradfitz/src/tailscale.com/types/logger/logger.go:31 +0x98
tailscale.com/netcheck.(*Client).logf(0xc0002a0000, 0x8b443b, 0x2, 0xc000280070, 0x1, 0x1)
        /home/bradfitz/src/tailscale.com/netcheck/netcheck.go:97 +0x65
tailscale.com/netcheck.(*Client).logConciseReport(0xc0002a0000, 0xc000390a20)
        /home/bradfitz/src/tailscale.com/netcheck/netcheck.go:559 +0x7f3
tailscale.com/netcheck.(*Client).GetReport(0xc0002a0000, 0x954460, 0xc000196f40, 0x0, 0x0, 0x0)
        /home/bradfitz/src/tailscale.com/netcheck/netcheck.go:487 +0xcf2
tailscale.com/wgengine/magicsock.(*Conn).updateNetInfo(0xc000298000, 0x9544e0, 0xc000192a80, 0x0, 0x0, 0x0)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:339 +0x163
tailscale.com/wgengine/magicsock.(*Conn).determineEndpoints(0xc000298000, 0x954460, 0xc000292080, 0x1, 0x1, 0xc000102420, 0xc00004ef00, 0x1, 0x0)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:503 +0x5d
tailscale.com/wgengine/magicsock.(*Conn).updateEndpoints(0xc000298000, 0x8b53ab, 0x7)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:302 +0x125
created by tailscale.com/wgengine/magicsock.(*Conn).ReSTUN
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1488 +0x1fb
FAIL    tailscale.com/wgengine  0.552s
FAIL
```

### PR DESCRIPTION
**Pull request by bradfitz**
all: make client use server-provided DERP map, add DERP region support

Instead of hard-coding the DERP map (except for cmd/tailscale netcheck
for now), get it from the control server at runtime.

And make the DERP map support multiple nodes per region with clients
picking the first one that's available. (The server will balance the
order presented to clients for load balancing)

This deletes the stunner package, merging it into the netcheck package
instead, to minimize all the config hooks that would've been
required.

Also fix some test flakes & races.

Fixes #387 (Don't hard-code the DERP map)
Updates #388 (Add DERP region support)
Fixes #399 (wgengine: flaky tests)

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/400)

### CODE REVIEW

File: derp/derphttp/derphttp_client.go [hunk 1]
```diff
@@ -243,6 +292,133 @@ func (c *Client) connect(ctx context.Context, caller string) (client *derp.Clien
...
+	return pred(ip)
+}
+
+func (c *Client) dialNode(ctx context.Context, n *tailcfg.DERPNode) (net.Conn, error) {
+	type res struct {
+		c   net.Conn
+		err error
+	}
+	resc := make(chan res) // must be unbuffered
+	ctx, cancel := context.WithTimeout(ctx, 1500*time.Millisecond)
```
Comment (crawshaw) [inline derp/derphttp/derphttp_client.go:367]: So nodes for a region are dialed in order, with a 1.5s timeout before trying the next node? Load balancing comes from the server in the ordering of the Nodes slice?

If so maybe put that in a comment in dialRegion.

File: derp/derpmap/derpmap.go [hunk 2]
```diff
@@ -7,151 +7,59 @@ package derpmap
...
-
-func (s *Server) String() string {
-	if s == nil {
-		return "<nil *derpmap.Server>"
-	}
-	if s.Geo != "" {
-		return fmt.Sprintf("%v (%v)", s.HostHTTPS, s.Geo)
+// Prod returns Tailscale's map of relay servers.
+//
+// This list is only used by cmd/tailscale's netcheck subcommand. In
```
Comment (crawshaw) [inline derp/derpmap/derpmap.go:40]: Should we persist the DERPMap somewhere so the netcheck command can use the latest copy? Another day.

File: netcheck/netcheck.go [hunk 3]
```diff
@@ -125,375 +133,541 @@ func (c *Client) ReceiveSTUNPacket(pkt []byte, src *net.UDPAddr) {
...
+			return false
+		}
+		return da < db
+	})
+	return prev
+}
 
-	const defaultMaxTries = 2
-	maxTries = map[string]int{}
+// makePlan generates the probe plan for a DERPMap, given the most
```
Comment (crawshaw) [inline netcheck/netcheck.go:236]: makeProbePlan

File: netcheck/netcheck.go [hunk 4]
```diff
@@ -125,375 +133,541 @@ func (c *Client) ReceiveSTUNPacket(pkt []byte, src *net.UDPAddr) {
...
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+
+	// If the probe is for a region we don't yet know about, that
+	// would help.
+	if _, ok := rs.report.RegionLatency[node.RegionID]; !ok {
+		return true
+	}
+
+	// If the probe is for IPv6 and we don't
```
Comment (crawshaw) [inline netcheck/netcheck.go:436]: sentence fell off a

File: wgengine/userspace.go [hunk 5]
```diff
@@ -49,7 +50,7 @@ const minimalMTU = 1280
 type userspaceEngine struct {
 	logf      logger.Logf
 	reqCh     chan struct{}
-	waitCh    chan struct{}
+	waitCh    chan struct{} // chan is closed when first Close call fully completes; contrast with closed bool
```
Comment (crawshaw) [inline wgengine/userspace.go:53]: closing bool?

File: derp/derpmap/derpmap.go (hunk 2 above)
Comment (bradfitz) [reply derp/derpmap/derpmap.go:40]: I think primary will be fetching from a new unauthenticated handler on our site.

If that fails, then we can try to STUN to other popular STUN servers (e.g. Google's) to see if UDP even works (cmd/tailscale netcheck is just a debugging tool for user, mostly). netcheck can also then report that HTTP or HTTPS is blocked.

But yeah, another day.

File: derp/derphttp/derphttp_client.go (hunk 1 above)
Comment (bradfitz) [reply derp/derphttp/derphttp_client.go:367]: Done.

**Pull request merged**